### ISSUE TITLE
alt is not a valid attribute for svg element

### ISSUE THREAD

**Issue by carmacleod** [carbon-design-system/carbon#1601]
`alt` is not a valid attribute for [svg element](https://svgwg.org/svg2-draft/struct.html#SVGElement).

Please search all of the carbon code (vanilla, react, ...) for places where there's an alt attribute on an svg element, and delete the alt attribute. There are some instances on the [Data Table example code page](http://www.carbondesignsystem.com/components/data-table/code).
I did not check other pages, but a code search would turn them up.  :)

Where appropriate, `aria-label` or `aria-labelledby` can be used on the svg (instead of alt), however first consider:
- is the alt text the empty string? i.e. `alt=""`? If so, add `aria-hidden="true"` to the svg (not aria-label).
- is there a visible label that would be redundant? If so, add `aria-hidden="true"`, as above, to make the svg invisible to screen readers.
- is the svg inside a button or link? If so, either put aria-label on the button/link and `aria-hidden="true"` on the svg, OR let the user know that it is an image button by adding `role="img"` to the svg and giving it an aria-label.
- is there a tooltip? if so:
  - for platform tooltip, add a title element as the first child of the svg, and use aria-labelledby to point to the unique id of the title element
  - for a custom tooltip, use aria-labelledby to point to the unique id of the custom tooltip's container element

**Comment by carmacleod** [comment]
Note also that creating a Carbon `<Icon description="Close" />` will create an SVG with `alt="Close"`.
Please look through the `<Icon>` code and make sure to remove all the `alt`'s from there as well. Thanks! :)

**Comment by joshblack** [comment]
Hi @carmacleod! 👋 Just for context into our icons, we're looking to migrate to `@carbon/icons-react` in a future release. The icons in that package will use `@carbon/icon-helpers` specifically for making sure we have the correct attributes on each `<svg>` element. You can see the specific method here: https://github.com/IBM/carbon-elements/blob/master/packages/icon-helpers/src/getAttributes.js

Would be great to have your thoughts on our logic there to make sure that we are setting the correct attributes 👍

**Comment by carmacleod** [comment]
Hi @joshblack !  I think the icon attribute logic is good! It pretty much follows some [little svg attribute patterns I wrote up](https://github.ibm.com/IBMa/Tools-Rules-HTML/issues/6#issuecomment-5156418) a while back.

The only thing I wonder about is custom tooltips, such as the [Carbon Icon Tooltip](https://www.carbondesignsystem.com/components/tooltip/code). (also, [react version](http://react.carbondesignsystem.com/?selectedKind=TooltipIcon&selectedStory=default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel))

I'm wondering why Carbon doesn't push custom tooltips more.  :)
The SVG title element (and the HTML title attribute) have a real drawback - they are [useless to sighted keyboard-only users](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title#Accessibility_concerns). (Also, apparently the svg title element isn't rendered as a tooltip by all browsers).

So, for example, in the [Carbon react Toolbar](http://react.carbondesignsystem.com/?selectedKind=Toolbar&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel), a sighted keyboard user would not need an explanation for the magnifying glass tool (clearly a search function) or the kebab tool (clearly a menu of some kind), but the teeny-tiny 'z' icon? What is that? No tooltip on focus... Most keyboarders would simply never try it.  :)

The advantage of custom tooltips is that they can be shown to sighted keyboard users, either:
- on focus of the element (usually a button or tool item), usually after a slight delay
- or on demand (when the element is focused) with the Shift+F1 key combination (see "Pop up tooltip for currently focused control" and "Show context-sensitive help for currently focused window or control" in [Wikipedia's Table of Keyboard Shortcuts](https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts)).

Anyhow, back to the icon attribute logic, if Carbon users are using custom tooltips, then you will want to have `role="img"` and either `aria-labelledby="id-of-tooltip-text-element"` or `aria-label="same string as tooltip text"` on the svg.

**Comment by carmacleod** [comment]
@joshblack Is the icon migration complete? i.e. there shouldn't be any "alt" attributes on svg's anywhere in Carbon Vanilla or React? If so, this issue can be closed.

Also, just wondering if you had a chance to give my last comment about recommending custom tooltips some thought? I can open a new issue for any discussion if that's preferred. :)

**Comment by joshblack** [comment]
Hi @carmacleod! 👋 We're still fine-tuning the developer experience, but the SVGs themselves should _not_ be setting `alt` and should be using one of the valid strategies outlined in: https://www.deque.com/blog/creating-accessible-svgs/

RE: tooltips, agreed 100%! Would love to get your thoughts on how this is done over in: https://ui.reach.tech/tooltip

**Comment by carmacleod** [comment]
> the SVGs themselves should not be setting alt

Nice - thanks! So this issue can be closed, and I will read through the tooltips link you sent and add thoughts there.  :)
